src/librustc_trans/back/link.rs

```diff
@@ -621,6 +621,67 @@ fn link_natively(sess: &Session, dylib: bool,
     let (pname, mut cmd) = get_linker(sess);
     cmd.env("PATH", command_path(sess));
 
+    let is_lld = pname.ends_with("lld");
+
+    if is_lld {
+        if sess.target.target.options.is_like_msvc {
+            cmd.args(&["-flavor", "link"]);
+        } else if sess.target.target.options.is_like_osx {
+            cmd.args(&["-flavor", "darwin"]);
+        } else {
+            cmd.args(&["-flavor", "gnu"]);
+        }
+    }
+
+    if sess.target.target.options.is_like_msvc {
+        let mut linker = MsvcLinker {
+            name: pname,
+            cmd: cmd,
+            sess: &sess,
+        };
+        link_natively_helper(&mut linker,
+                             sess,
+                             dylib,
+                             objects,
+                             out_filename,
+                             trans,
+                             outputs,
+                             tmpdir);
+    } else {
+        let mut linker = GnuLinker {
+            name: pname,
+            cmd: cmd,
+            sess: &sess,
+            is_lld: is_lld,
+        };
+        link_natively_helper(&mut linker,
+                             sess,
+                             dylib,
+                             objects,
+                             out_filename,
+                             trans,
+                             outputs,
+                             tmpdir);
+    }
+
+    // On OSX, debuggers need this utility to get run to do some munging of
+    // the symbols
+    if sess.target.target.options.is_like_osx && sess.opts.debuginfo != NoDebugInfo {
+        match Command::new("dsymutil").arg(out_filename).output() {
+            Ok(..) => {}
+            Err(e) => sess.fatal(&format!("failed to run dsymutil: {}", e)),
+        }
+    }
+}
+
+fn link_natively_helper<T: Linker>(cmd: &mut T,
```

Comment:
The proximate cause of this having to be generic was the generic arguments to `Linker.arg{,s}`.

This function was split out because I needed to put the bound on something.